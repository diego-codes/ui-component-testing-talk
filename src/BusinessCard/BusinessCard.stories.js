import React from 'react';
import { storiesOf } from '@storybook/react';

import BusinessCard from '.';

const stories = storiesOf(BusinessCard.name, module);

stories.add('default', () => (
  <BusinessCard
    firstName="Squidward"
    lastName="Tentacles"
    title="Cashier &amp; Clarinet master"
    business="Krusty Krab"
  />
));

stories.add('online', () => (
  <BusinessCard
    image="https://pbs.twimg.com/profile_images/439113071902998528/H3CCnV1C.jpeg"
    firstName="Eugene"
    lastName="Krabs"
    title="Restauranteur"
    business="Krusty Krab"
    online
  />
));

stories.add('online with phone number', () => (
  <BusinessCard
    image="http://en.spongepedia.org/images/thumb/5/55/SpongeBob_SquarePants_Sheldon_Plankton.jpg/250px-SpongeBob_SquarePants_Sheldon_Plankton.jpg"
    firstName="Sheldon"
    lastName="Plankton"
    title="Evil genius"
    business="Chum Bucket"
    phone="555-555-5555"
    online
  />
));

stories.add('offline with email', () => (
  <BusinessCard
    image="https://pbs.twimg.com/profile_images/532352938392293376/15Zj9aOq_400x400.png"
    firstName="Sandy"
    lastName="Cheeks"
    title="Scientist"
    business="Independent"
    email="sandy.cheeks@scientist.sea"
  />
));

stories.add('online with contact info', () => (
  <BusinessCard
    image="https://pbs.twimg.com/profile_images/1856175966/Spongebob_400x400.jpg"
    firstName="SpongeBob"
    lastName="SquarePants"
    title="Fry cook"
    business="Krusty Krab"
    email="s.sqarepants@krabby.com"
    phone="555-555-5555"
  />
));
