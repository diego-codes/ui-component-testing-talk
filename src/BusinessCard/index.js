import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { OverflowMenu, OverflowMenuItem } from '../OverflowMenu';
import {
  Container,
  ContactInfo,
  HeadShot,
  FullName,
  Subtitle,
  Footer,
  OnlineStatus,
} from './styles';

const BusinessCard = ({
  firstName,
  lastName,
  title,
  business,
  online,
  image,
  phone,
  email,
  onChat,
  onVoiceCall,
  onVideoCall,
  onSMS,
  onEmail,
}) => (
  <Container>
    <ContactInfo>
      {image && <HeadShot image={image} data-testid="business-card-image" />}
      <div>
        <FullName>{`${firstName} ${lastName}`}</FullName>
        <Subtitle>{`${title} â€” ${business}`}</Subtitle>
      </div>
    </ContactInfo>
    <Footer>
      <OnlineStatus online={online}>{online ? 'Online' : 'Offline'}</OnlineStatus>
      {(online || phone || email) && (
        <OverflowMenu toggleText={`Contact ${firstName}`} flipped>
          {online && <OverflowMenuItem onClick={onChat}>Start chat</OverflowMenuItem>}
          {phone && (
            <Fragment>
              <OverflowMenuItem onClick={onVoiceCall}>Start voice call</OverflowMenuItem>
              <OverflowMenuItem onClick={onVideoCall}>Start video call</OverflowMenuItem>
              <OverflowMenuItem onClick={onSMS}>Send SMS</OverflowMenuItem>
            </Fragment>
          )}
          {email && <OverflowMenuItem onClick={onEmail}>Send email</OverflowMenuItem>}
        </OverflowMenu>
      )}
    </Footer>
  </Container>
);

BusinessCard.propTypes = {
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  business: PropTypes.string.isRequired,
  online: PropTypes.bool,
  image: PropTypes.string,
  phone: PropTypes.string,
  email: PropTypes.string,
  onChat: PropTypes.func,
  onVoiceCall: PropTypes.func,
  onVideoCall: PropTypes.func,
  onSMS: PropTypes.func,
  onEmail: PropTypes.func,
};

BusinessCard.defaultProps = {
  online: false,
  image: undefined,
  phone: undefined,
  email: undefined,
  onChat: () => {},
  onVoiceCall: () => {},
  onVideoCall: () => {},
  onSMS: () => {},
  onEmail: () => {},
};

export default BusinessCard;
