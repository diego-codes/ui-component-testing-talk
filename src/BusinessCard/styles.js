import styled from 'styled-components';

export const Container = styled.div`
  font-family: 'IBM Plex Sans';
  font-size: 0.875rem;
  line-height: 1.2;
  background-color: #f3f3f3;
  color: #252525;
  max-width: 32em;
`;

export const ContactInfo = styled.div`
  padding: 2em;
  display: flex;
  align-items: center;
`;

export const HeadShot = styled.div`
  width: 5em;
  height: 5em;
  display: inline-block;
  background-image: url(${props => props.image});
  background-size: cover;
  background-position: center;
  background-repeat: none;
  border-radius: 4px;
  margin-right: 1em;
`;

export const FullName = styled.h1`
  font-size: 2em;
  font-weight: 600;
  letter-spacing: -0.03em;
  margin: 0;
  margin-bottom: 0.1em;
`;

export const Subtitle = styled.h2`
  margin: 0;
  font-size: 1em;
  font-weight: 400;
  color: #3d3d3d;
`;

export const Footer = styled.footer`
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #3d3d3d;
  font-size: 0.9em;
  height: 2em;
  color: #f3f3f3;
`;

export const OnlineStatus = styled.span`
  display: inline-flex;
  align-items: center;
  padding: 0 2em;

  &:before {
    content: '';
    display: inline-block;
    height: 0.5em;
    width: 0.5em;
    border-radius: 100%;
    background-color: ${props => (props.online ? '#56D679' : 'transparent')};
    box-shadow: inset 0 0 0 1px ${props => (props.online ? 'transparent' : 'currentColor')};
    margin-right: 0.4em;
  }
`;
