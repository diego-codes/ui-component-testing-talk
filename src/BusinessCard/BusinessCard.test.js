import React from 'react';
import { render, queryByTestId, queryAllByText } from 'react-testing-library';
import BusinessCard from '.';

const user = {
  firstName: 'SpongeBob',
  lastName: 'SquarePants',
  title: 'Fry cook',
  business: 'Krusty Krab',
};

const image = 'spongebob.jpg';
const email = 's.squarepants@krabby.com';
const phone = '555-555-5555';

describe(BusinessCard.name, () => {
  describe('structure - detailed', () => {
    it('renders', () =>
      expect(() =>
        render(
          <BusinessCard
            firstName={user.firstName}
            lastName={user.lastName}
            title={user.title}
            business={user.business}
          />
        )
      ).not.toThrow());

    it('contains full name', () => {
      const { container } = render(<BusinessCard {...user} />);
      expect(container).toHaveTextContent(`${user.firstName} ${user.lastName}`);
    });

    it('displays headshot', () => {
      const { getByTestId } = render(<BusinessCard {...user} image={image} />);
      const headShot = getByTestId('business-card-image');
      expect(headShot).not.toBeNull();
    });

    it('doesn not display headshot if image is not provided', () => {
      const { container } = render(<BusinessCard {...user} />);
      const headShot = queryByTestId(container, 'business-card-image');
      expect(headShot).toBeNull();
    });

    it('shows that user is online', () => {
      const { container } = render(<BusinessCard {...user} online />);
      expect(container).toHaveTextContent(/online/i);
    });

    it('does not contain contact menu when contact info is not provided', () => {
      const { container } = render(<BusinessCard {...user} />);
      const menuButton = queryByTestId(container, 'overflow-menu');
      expect(menuButton).toBeNull();
    });

    it('contains contact menu when phone number is provided', () => {
      const { getByTestId } = render(<BusinessCard {...user} phone={phone} />);
      const menuButton = getByTestId('overflow-menu');
      expect(menuButton).not.toBeNull();
    });

    it('contains contact menu when email is provided', () => {
      const { getByTestId } = render(<BusinessCard {...user} email={email} />);
      const menuButton = getByTestId('overflow-menu');
      expect(menuButton).not.toBeNull();
    });

    it('contains phone menu options when phone number is provided', () => {
      const { getByTestId } = render(<BusinessCard {...user} phone={phone} />);
      const overflowMenu = getByTestId('overflow-menu');
      const overflowMenuToggleButton = getByTestId('overflow-menu-toggle');
      overflowMenuToggleButton.click();
      const options = queryAllByText(overflowMenu, /call|sms/i);
      expect(options).toHaveLength(3);
    });

    it('contains email menu options when email is provided', () => {
      const { getByTestId } = render(<BusinessCard {...user} email={email} />);
      const overflowMenu = getByTestId('overflow-menu');
      const overflowMenuToggleButton = getByTestId('overflow-menu-toggle');
      overflowMenuToggleButton.click();
      const options = queryAllByText(overflowMenu, /email/i);
      expect(options).toHaveLength(1);
    });
  });

  describe('Structure - Snapshots', () => {
    it('renders head shot', () => {
      const { container } = render(<BusinessCard {...user} image={image} />);
      expect(container).toMatchSnapshot();
    });

    it('renders offline indicator', () => {
      const { container } = render(<BusinessCard {...user} />);
      expect(container).toMatchSnapshot();
    });

    it('renders online indicator', () => {
      const { container } = render(<BusinessCard {...user} online />);
      expect(container).toMatchSnapshot();
    });

    it('renders contact menu when phone number is provided', () => {
      const { container } = render(<BusinessCard {...user} phone={phone} />);
      expect(container).toMatchSnapshot();
    });

    it('renders contact menu when email is provided', () => {
      const { container } = render(<BusinessCard {...user} email={email} />);
      expect(container).toMatchSnapshot();
    });

    it('renders phone number menu options when phone number is provided', () => {
      const { container, getByTestId } = render(<BusinessCard {...user} phone={phone} />);
      getByTestId('overflow-menu-toggle').click();
      expect(container).toMatchSnapshot();
    });

    it('renders email menu options when email is provided', () => {
      const { container, getByTestId } = render(<BusinessCard {...user} email={email} />);
      getByTestId('overflow-menu-toggle').click();
      expect(container).toMatchSnapshot();
    });
  });

  describe('Behavior', () => {
    it('invokes onVoiceCall', () => {
      const voiceCallMock = jest.fn();
      const { getByTestId, getByText } = render(
        <BusinessCard {...user} phone={phone} onVoiceCall={voiceCallMock} />
      );
      getByTestId('overflow-menu-toggle').click();
      getByText(/voice call/i).click();
      expect(voiceCallMock).toHaveBeenCalled();
    });

    it('invokes onVideoCall', () => {
      const videoCallMock = jest.fn();
      const { getByTestId, getByText } = render(
        <BusinessCard {...user} phone={phone} onVideoCall={videoCallMock} />
      );
      getByTestId('overflow-menu-toggle').click();
      getByText(/video call/i).click();
      expect(videoCallMock).toHaveBeenCalledTimes(1);
    });

    it('invokes onSMS', () => {
      const sendSMSMock = jest.fn();
      const { getByTestId, getByText } = render(
        <BusinessCard {...user} phone={phone} onSMS={sendSMSMock} />
      );
      getByTestId('overflow-menu-toggle').click();
      getByText(/SMS/i).click();
      expect(sendSMSMock).toHaveBeenCalledTimes(1);
    });

    it('invokes onEmail', () => {
      const sendEmail = jest.fn();
      const { getByTestId, getByText } = render(
        <BusinessCard {...user} email={email} onEmail={sendEmail} />
      );
      getByTestId('overflow-menu-toggle').click();
      getByText(/email/i).click();
      expect(sendEmail).toHaveBeenCalledTimes(1);
    });
  });
});
