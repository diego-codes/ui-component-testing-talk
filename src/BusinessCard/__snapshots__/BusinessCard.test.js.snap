// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessCard Structure - Snapshots renders chat menu options when contact is online 1`] = `
<div>
  <div
    class="sc-EHOje jdCcvM"
  >
    <div
      class="sc-bZQynM hzprsO"
    >
      <div>
        <h1
          class="sc-htoDjs iZtlym"
        >
          SpongeBob SquarePants
        </h1>
        <h2
          class="sc-dnqmqq kGYEvA"
        >
          Fry cook — Krusty Krab
        </h2>
      </div>
    </div>
    <footer
      class="sc-iwsKbI krTmKM"
    >
      <span
        class="sc-gZMcBi hQsnUx"
        data-testid="business-card-online-indicator"
      >
        Online
      </span>
      <div
        class="sc-bdVaJa huqire"
        data-testid="overflow-menu"
      >
        <button
          class="sc-bwzfXH sc-htpNat tfKyv"
          data-testid="overflow-menu-toggle"
          open=""
        >
          Contact SpongeBob
        </button>
        <div
          class="sc-ifAKCX gwmupT"
        >
          <button
            class="sc-bwzfXH sc-bxivhb cCHJQU"
          >
            Start chat
          </button>
        </div>
      </div>
    </footer>
  </div>
</div>
`;

exports[`BusinessCard Structure - Snapshots renders contact menu when email is provided 1`] = `
<div>
  <div
    class="sc-EHOje jdCcvM"
  >
    <div
      class="sc-bZQynM hzprsO"
    >
      <div>
        <h1
          class="sc-htoDjs iZtlym"
        >
          SpongeBob SquarePants
        </h1>
        <h2
          class="sc-dnqmqq kGYEvA"
        >
          Fry cook — Krusty Krab
        </h2>
      </div>
    </div>
    <footer
      class="sc-iwsKbI krTmKM"
    >
      <span
        class="sc-gZMcBi kIVahm"
        data-testid="business-card-online-indicator"
      >
        Offline
      </span>
      <div
        class="sc-bdVaJa huqire"
        data-testid="overflow-menu"
      >
        <button
          class="sc-bwzfXH sc-htpNat gTrUOb"
          data-testid="overflow-menu-toggle"
        >
          Contact SpongeBob
        </button>
      </div>
    </footer>
  </div>
</div>
`;

exports[`BusinessCard Structure - Snapshots renders contact menu when phone number is provided 1`] = `
<div>
  <div
    class="sc-EHOje jdCcvM"
  >
    <div
      class="sc-bZQynM hzprsO"
    >
      <div>
        <h1
          class="sc-htoDjs iZtlym"
        >
          SpongeBob SquarePants
        </h1>
        <h2
          class="sc-dnqmqq kGYEvA"
        >
          Fry cook — Krusty Krab
        </h2>
      </div>
    </div>
    <footer
      class="sc-iwsKbI krTmKM"
    >
      <span
        class="sc-gZMcBi kIVahm"
        data-testid="business-card-online-indicator"
      >
        Offline
      </span>
      <div
        class="sc-bdVaJa huqire"
        data-testid="overflow-menu"
      >
        <button
          class="sc-bwzfXH sc-htpNat gTrUOb"
          data-testid="overflow-menu-toggle"
        >
          Contact SpongeBob
        </button>
      </div>
    </footer>
  </div>
</div>
`;

exports[`BusinessCard Structure - Snapshots renders default structure 1`] = `
<div>
  <div
    class="sc-EHOje jdCcvM"
  >
    <div
      class="sc-bZQynM hzprsO"
    >
      <div>
        <h1
          class="sc-htoDjs iZtlym"
        >
          SpongeBob SquarePants
        </h1>
        <h2
          class="sc-dnqmqq kGYEvA"
        >
          Fry cook — Krusty Krab
        </h2>
      </div>
    </div>
    <footer
      class="sc-iwsKbI krTmKM"
    >
      <span
        class="sc-gZMcBi kIVahm"
        data-testid="business-card-online-indicator"
      >
        Offline
      </span>
    </footer>
  </div>
</div>
`;

exports[`BusinessCard Structure - Snapshots renders email menu options when email is provided 1`] = `
<div>
  <div
    class="sc-EHOje jdCcvM"
  >
    <div
      class="sc-bZQynM hzprsO"
    >
      <div>
        <h1
          class="sc-htoDjs iZtlym"
        >
          SpongeBob SquarePants
        </h1>
        <h2
          class="sc-dnqmqq kGYEvA"
        >
          Fry cook — Krusty Krab
        </h2>
      </div>
    </div>
    <footer
      class="sc-iwsKbI krTmKM"
    >
      <span
        class="sc-gZMcBi kIVahm"
        data-testid="business-card-online-indicator"
      >
        Offline
      </span>
      <div
        class="sc-bdVaJa huqire"
        data-testid="overflow-menu"
      >
        <button
          class="sc-bwzfXH sc-htpNat tfKyv"
          data-testid="overflow-menu-toggle"
          open=""
        >
          Contact SpongeBob
        </button>
        <div
          class="sc-ifAKCX gwmupT"
        >
          <button
            class="sc-bwzfXH sc-bxivhb cCHJQU"
          >
            Send email
          </button>
        </div>
      </div>
    </footer>
  </div>
</div>
`;

exports[`BusinessCard Structure - Snapshots renders head shot 1`] = `
<div>
  <div
    class="sc-EHOje jdCcvM"
  >
    <div
      class="sc-bZQynM hzprsO"
    >
      <div
        class="sc-gzVnrw kguMlY"
        data-testid="business-card-image"
        style="background-image: url(spongebob.jpg);"
      />
      <div>
        <h1
          class="sc-htoDjs iZtlym"
        >
          SpongeBob SquarePants
        </h1>
        <h2
          class="sc-dnqmqq kGYEvA"
        >
          Fry cook — Krusty Krab
        </h2>
      </div>
    </div>
    <footer
      class="sc-iwsKbI krTmKM"
    >
      <span
        class="sc-gZMcBi kIVahm"
        data-testid="business-card-online-indicator"
      >
        Offline
      </span>
    </footer>
  </div>
</div>
`;

exports[`BusinessCard Structure - Snapshots renders online indicator and contact menu 1`] = `
<div>
  <div
    class="sc-EHOje jdCcvM"
  >
    <div
      class="sc-bZQynM hzprsO"
    >
      <div>
        <h1
          class="sc-htoDjs iZtlym"
        >
          SpongeBob SquarePants
        </h1>
        <h2
          class="sc-dnqmqq kGYEvA"
        >
          Fry cook — Krusty Krab
        </h2>
      </div>
    </div>
    <footer
      class="sc-iwsKbI krTmKM"
    >
      <span
        class="sc-gZMcBi hQsnUx"
        data-testid="business-card-online-indicator"
      >
        Online
      </span>
      <div
        class="sc-bdVaJa huqire"
        data-testid="overflow-menu"
      >
        <button
          class="sc-bwzfXH sc-htpNat gTrUOb"
          data-testid="overflow-menu-toggle"
        >
          Contact SpongeBob
        </button>
      </div>
    </footer>
  </div>
</div>
`;

exports[`BusinessCard Structure - Snapshots renders phone number menu options when phone number is provided 1`] = `
<div>
  <div
    class="sc-EHOje jdCcvM"
  >
    <div
      class="sc-bZQynM hzprsO"
    >
      <div>
        <h1
          class="sc-htoDjs iZtlym"
        >
          SpongeBob SquarePants
        </h1>
        <h2
          class="sc-dnqmqq kGYEvA"
        >
          Fry cook — Krusty Krab
        </h2>
      </div>
    </div>
    <footer
      class="sc-iwsKbI krTmKM"
    >
      <span
        class="sc-gZMcBi kIVahm"
        data-testid="business-card-online-indicator"
      >
        Offline
      </span>
      <div
        class="sc-bdVaJa huqire"
        data-testid="overflow-menu"
      >
        <button
          class="sc-bwzfXH sc-htpNat tfKyv"
          data-testid="overflow-menu-toggle"
          open=""
        >
          Contact SpongeBob
        </button>
        <div
          class="sc-ifAKCX gwmupT"
        >
          <button
            class="sc-bwzfXH sc-bxivhb cCHJQU"
          >
            Start voice call
          </button>
          <button
            class="sc-bwzfXH sc-bxivhb cCHJQU"
          >
            Start video call
          </button>
          <button
            class="sc-bwzfXH sc-bxivhb cCHJQU"
          >
            Send SMS
          </button>
        </div>
      </div>
    </footer>
  </div>
</div>
`;
